src/internal/storage/fileset/prefetcher_test.go

```diff
@@ -0,0 +1,101 @@
+package fileset
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"math/rand"
+	"strconv"
+	"testing"
+
+	units "github.com/docker/go-units"
+	"github.com/pachyderm/pachyderm/v2/src/internal/randutil"
+	"github.com/pachyderm/pachyderm/v2/src/internal/require"
+)
+
+func createTestFileSet(tb testing.TB, ctx context.Context, s *Storage, numFiles int, fileSize int, random *rand.Rand) FileSet {
+	w := s.NewWriter(ctx)
+	for i := 0; i < numFiles; i++ {
+		data := randutil.Bytes(random, fileSize)
+		require.NoError(tb, w.Add(fmt.Sprintf("%08d", i), DefaultFileDatum, bytes.NewReader(data)))
+	}
+	id, err := w.Close()
+	require.NoError(tb, err)
+	fs, err := s.Open(ctx, []ID{*id})
+	require.NoError(tb, err)
+	return fs
+}
+
+func TestPrefetcher(t *testing.T) {
+	ctx := context.Background()
+	s := newTestStorage(t)
+	seed := int64(1648577872380609229)
+	random := rand.New(rand.NewSource(seed))
+	tests := []struct {
+		name     string
+		numFiles int
+		fileSize int
+	}{
+		{"0B", 1000000, 0},
+		{"1KB", 100000, units.KB},
+		{"100KB", 1000, 100 * units.KB},
+		{"10MB", 10, 10 * units.MB},
+		{"100MB", 1, 100 * units.MB},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			fs := createTestFileSet(t, ctx, s, test.numFiles, test.fileSize, random)
+			fs = NewPrefetcher(s, fs)
+			var i int
+			require.NoError(t, fs.Iterate(ctx, func(f File) error {
+				p, err := strconv.Atoi(f.Index().Path)
+				require.NoError(t, err)
+				require.Equal(t, i, p)
+				i++
+				return nil
+			}))
+			require.Equal(t, test.numFiles, i)
+		})
+	}
+}
+
+func BenchmarkNoPrefetcher(b *testing.B) {
```

Comment:
Glad you included these benchmarks. Locally it took about 2.5 minutes without the prefetcher vs 30 seconds with it. Big difference!